fly/integration/execute_test.go

```diff
@@ -246,6 +247,13 @@ run:
 			atcServer.RouteToHandler("POST", "/api/v1/teams/main/pipelines/some-pipeline/builds",
 				ghttp.CombineHandlers(
 					ghttp.VerifyRequest("POST", "/api/v1/teams/main/pipelines/some-pipeline/builds"),
+					testhelpers.VerifyResourceTypes(atc.VersionedResourceType{
```

Comment:
I'm wondering if we could test the addition of `VersionedResourceTypes` to the execute task plan by testing that the `atc.TaskPlan` contains the proper `atc.VersionedResourceTypes`? So we would add
`VerifyPlan(expectedPlan)` with the expectedPlan contain the proper `VersionedResourceTypes` rather than adding `VerifyResourceTypes(..`, and that would allow us to avoid adding this new resource types matcher. Let me know what you think!

Reply:
I started writing the new matcher because I was being lazy and didn't want to figure out the structure of the pipeline in the test. It also felt brittle to break the test, for example, when the test fixture changes, because the expected pipeline didn't match. That said, I'm happy to get rid of the matcher if you do not want to maintain yet another matcher. I don't have a strong opinion.

Reply:
Looking at this again, it looks like all the other tests share the same `expectedPlan`, so I could add the `VersionedResourceType` to the task config in the `BeforeEach`. I am still worried about the goal of the test not being clear over time. For example if a new Task or Get is added, the author of the change could simply modify the top-level `expectedPlan` to add the new task, without `VersionedResourceTypes`. This change would make this test trivially pass.

Reply:
I see, I definitely understand your concern. I think preferably we would want to avoid adding another matcher but if you feel like adding a resource type matcher would be the best method to testing this, then I would be for it.

I tried seeing what it would look like if we were to test using the `expectedPlan`, I'd love to get your opinion on it! The first thing I did was extracting out the `atc.TaskPlan` from the `expectedPlan`

```
taskPlan = planFactory.NewPlan(atc.TaskPlan{
	Name: "one-off",
	Config: &atc.TaskConfig{
		Platform: "some-platform",
		ImageResource: &atc.ImageResource{
			Type: "registry-image",
			Source: atc.Source{
				"repository": "ubuntu",
			},
		},
		Inputs: []atc.TaskInputConfig{
			{Name: "fixture"},
		},
		Params: map[string]string{
			"FOO":   "bar",
			"BAZ":   "buzz",
			"X":     "1",
			"EMPTY": "",
		},
		Run: atc.TaskRunConfig{
			Path: "find",
			Args: []string{"."},
		},
	},
})

expectedPlan = planFactory.NewPlan(atc.DoPlan{
	planFactory.NewPlan(atc.AggregatePlan{
		planFactory.NewPlan(atc.ArtifactInputPlan{
			ArtifactID: 125,
			Name:       filepath.Base(buildDir),
		}),
	}),
	taskPlan,
})
```

And then modifying the `VersionedResourceTypes` within the `Task` for the `taskPlan`.

```
Context("when there is a pipeline job with the same input", func() {
	BeforeEach(func() {
		taskPlan.Task.VersionedResourceTypes = atc.VersionedResourceTypes{
			atc.VersionedResourceType{
				ResourceType: atc.ResourceType{
					Name:   "resource-type",
					Type:   "s3",
					Source: atc.Source{},
				},
			},
		}
```

Then we would verify the plan is the same as the expectedPlan

```
atcServer.RouteToHandler("POST", "/api/v1/teams/main/pipelines/some-pipeline/builds",
	ghttp.CombineHandlers(
		ghttp.VerifyRequest("POST", "/api/v1/teams/main/pipelines/some-pipeline/builds"),
		VerifyPlan(expectedPlan),
		func(w http.ResponseWriter, r *http.Request) {
			http.SetCookie(w, &http.Cookie{
				Name:    "Some-Cookie",
				Value:   "some-cookie-data",
				Path:    "/",
				Expires: time.Now().Add(1 * time.Minute),
			})
		},
		ghttp.RespondWith(201, `{"id":128}`),
	),
)
```

Originally that would've been enough but because we are now verifying the build plan within the `"when there is a pipeline job with the same input"` test, it also has an extra change that adds build inputs to the plan so we would need to have the expectedPlan within the `BeforeEach` of the test respect that build inputs modification. I think this actually adds value to this test because it covers this [part of the codebase](https://github.com/concourse/concourse/blob/d64843ce1bbab2309213ff677e25aadb950fc33b/fly/commands/internal/executehelpers/inputs.go#L46-L89).

```
expectedPlan = planFactory.NewPlan(atc.DoPlan{
	planFactory.NewPlan(atc.AggregatePlan{
		planFactory.NewPlan(atc.GetPlan{
			Name: "fixture",
			VersionedResourceTypes: atc.VersionedResourceTypes{
				atc.VersionedResourceType{
					ResourceType: atc.ResourceType{
						Name:   "resource-type",
						Type:   "s3",
						Source: atc.Source{},
					},
				},
			},
		}),
	}),
	taskPlan,
})
```

Reply:
I don't have a strong opinion either way. I will make the changes proposed above tomorrow morning in a separate commit.

Reply:
I pushed the changes requested above.